Simplification with assumptions: refine and abs

maurogaravello opened the pull request:
This deals with  #12724. In case of `refine(abs(x*y), Q.positive(x))`
the result was `Abs(x*y)`, which is mathematically correct. A better result
is `x * Abs(y)`.

This PR proposes a change in `refine_abs()` method in `sympy/assumptions/refine.py`.
In the case the arg is a multiplication, the `Abs` is splitted according to the factors of arg.
A test is added in `sympy/assumptions/tests/test_refine.py`.

maurogaravello committed:
Simplification with assumptions: refine and abs

This deals with issue 12724. In case of 'refine(abs(x*y), Q.positive(x))'
the result was 'Abs(x*y)', which is mathematically correct. A better result
is 'x * Abs(y)'.

This PR proposes a change in 'refine_abs()' method in 'sympy/assumptions/refine.py'.
In the case the arg is a multiplication, the Abs is splitted according to the factors of arg.
A test is added in 'sympy/assumptions/tests/test_refine.py'.

smichr commented on sympy/assumptions/refine.py:
```diff
@@ -70,6 +70,9 @@ def refine_abs(expr, assumptions):
         return -arg
+    # arg is Mul
+    if isinstance(arg, Mul):
+        return refine(abs(arg.args[0]), assumptions) * refine(abs(arg / arg.args[0]), assumptions)
```
Might we keep those that don't simplify out of Abs *in* the Abs? e.g. `Abs(x, y, z) -> x * Abs(y*z)` if only x is positive?

```
If isinstance(arg, Mul):
    r = [refine(abs(a), assumptions) for a in arg.args]
    nonabs = []
    inabs = []
    for i in r:
        if isinstance(i, Abs):
            inabs.append(i.args[0])
        else:
            nonabs.append(i)
    return Mul(*nonabs)*Abs(Mul(*inabs))
```

maurogaravello replied:
I agree with your  suggestion.
Probably today I will not able to push the commit.

maurogaravello committed:
Improving of refine_abs() method.

This commit follows the suggestion of keeping terms that don't simplify
inside tha Abs.
A test is modified accordingly.

smichr commented:
Thanks -- it's in.

maurogaravello commented:
Thanks
